hive/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveTableUtils.java

```diff
@@ -50,33 +56,48 @@ static Schema getSchema(org.apache.hadoop.hive.metastore.api.Table table) {
     Schema schema;
     if (schemaStr == null) {
       LOG.warn("Table {}.{} does not have an avro.schema.literal set; using Hive schema instead. The schema will not" +
-          " have case sensitivity and nullability information", table.getDbName(), table.getTableName());
+                   " have case sensitivity and nullability information", table.getDbName(), table.getTableName());
       // TODO: Add support for tables without avro.schema.literal
       throw new UnsupportedOperationException("Reading tables without avro.schema.literal not implemented yet");
     } else {
       schema = AvroSchemaUtil.toIceberg(new org.apache.avro.Schema.Parser().parse(schemaStr));
     }
 
-    List<String> partCols = table.getPartitionKeys().stream().map(FieldSchema::getName).collect(Collectors.toList());
-    return addPartitionColumnsIfRequired(schema, partCols);
-  }
+    Types.StructType dataStructType = schema.asStruct();
+    List<Types.NestedField> fields = Lists.newArrayList(dataStructType.fields());
 
-  private static Schema addPartitionColumnsIfRequired(Schema schema, List<String> partitionColumns) {
-    List<Types.NestedField> fields = new ArrayList<>(schema.columns());
-    AtomicInteger fieldId = new AtomicInteger(10000);
-    partitionColumns.stream().forEachOrdered(column -> {
-      Types.NestedField field = schema.findField(column);
-      if (field == null) {
-        // TODO: Support partition fields with non-string types
-        fields.add(Types.NestedField.required(fieldId.incrementAndGet(), column, Types.StringType.get()));
-      } else {
-        Preconditions.checkArgument(field.type().equals(Types.StringType.get()),
-            "Tables with non-string partition columns not supported yet");
+    Schema partitionSchema = partitionSchema(table.getPartitionKeys());
+    Types.StructType partitionStructType = partitionSchema.asStruct();
+    final Integer maxId = Collections.max(TypeUtil.indexById(dataStructType).keySet());
+    final Type partitionType = TypeUtil.assignFreshIds(partitionStructType, new TypeUtil.NextID() {
+      private int id = maxId;
+
+      @Override
+      public int get() {
+        id += 1;
+        return id;
       }
     });
+
+    fields.addAll(partitionType.asStructType().fields());
     return new Schema(fields);
   }
 
+  private static Schema partitionSchema(List<FieldSchema> partitionKeys) {
+    AtomicInteger fieldId = new AtomicInteger(10000);
+    List<Types.NestedField> partitionFields = partitionKeys
+        .stream()
+        .map(f -> Types.NestedField.optional(
+            fieldId.incrementAndGet(), f.getName(), icebergType(f.getType()), f.getComment()))
+        .collect(Collectors.toList());
+    return new Schema(partitionFields);
+  }
+
+  private static Type icebergType(String hiveTypeString) {
+    PrimitiveTypeInfo primitiveTypeInfo = TypeInfoFactory.getPrimitiveTypeInfo(hiveTypeString);
+    return new HiveTypeToIcebergType().primitive(primitiveTypeInfo);
```

Comment:
Call HiveTypeUtil.visit instead? Since that is the standard way to call these visitors.